### PR TITLE
wxgui: startup GUI automatic detection of grassdata: make case independent #644

### PR DESCRIPTION
**Pull request by lindakladivova**
After starting GRASS GIS with GUI, the GUI now searches for existing grassdata (e.g., dac6d4a and #644 with #664). 

This platfrom-dependent directory would be:

    - $HOME (os.path.expanduser('~')) on Linux and the like,
    - User's Documents on Windows (see dac6d4a for code trying to identify that dir),
    - One of the above on macOS - macOS users, please share your ideas.

This function does not address temporary directory. No platform specific code.

### CODE REVIEW

File: gui/wxpython/startup/utils.py [hunk 1]
```diff
@@ -35,24 +35,29 @@ def get_possible_database_path():
...
-        # just ignore the errors if it doesn't work
-        pass
-    path = None
-    for candidate in candidates:
-        if os.path.exists(candidate):
-            path = candidate
-            break  # get the first match
-    return path
+    
+    options = ["grassdata","GRASSDATA","Grassdata","GrassData"]
```
Comment (wenzeslaus) [inline gui/wxpython/startup/utils.py:39]: Please use PEP8 for formatting. In general, you can run Flake8 (`flake8 startup/utils.py`) to get some errors, but some are unfortunately still ignored/disabled including this one. Here, just remember _space after comma_ and use: `..."grassdata", "GRASSDATA"...`.

**Comment by mlennert** [comment]
Wouldn't it be easier to just check whether the string transformed to all minor case is equal to 'grassdata' ?

**Comment by lindakladivova** [comment]
> 
> 
> Wouldn't it be easier to just check whether the string transformed to all minor case is equal to 'grassdata' ?



> 
> 
> Wouldn't it be easier to just check whether the string transformed to all minor case is equal to 'grassdata' ?

I thought about it too, but the problem is that I don't have any string that I can transform. I have no idea what combination of lowercase and uppercase letters in the "grassdata" a user could have on his computer.

**Comment by neteler** [comment]
> I thought about it too, but the problem is that I don't have any string that I can transform. I have no idea what combination of lowercase and uppercase letters in the "grassdata" a user could have on his computer.

Maybe just convert all to lowercase?

https://grass.osgeo.org/programming7/strings_8c.html
--> `G_str_to_lower (char *str)`

```
It seems to be wrapped:
ag --python G_str_to_lower lib/python/ctypes/OBJ.x86_64-pc-linux-gnu/gis.py 
3834:if hasattr(_libs['grass_gis.7.9'], 'G_str_to_lower'):
3835:    G_str_to_lower = _libs['grass_gis.7.9'].G_str_to_lower
3836:    G_str_to_lower.argtypes = [String]
3837:    G_str_to_lower.restype = None
```

**Comment by mlennert** [comment]
> > Wouldn't it be easier to just check whether the string transformed to all minor case is equal to 'grassdata' ?
> 
> I thought about it too, but the problem is that I don't have any string that I can transform. I have no idea what combination of lowercase and uppercase letters in the "grassdata" a user could have on his computer.

I understand. You would have to approach it differently, i.e. get the list of all directories and then identify those that contain 'grassdata'. Something like this:
```
import os
for dir in next(os.walk('.'))[1]:
    if 'grassdata' in dir.lower():
        print(dir)
```
Obviously something else than print() is needed...

**Comment by hellik** [comment]
See e.g. 

https://www.itprotoday.com/cloud-computing/what-environment-variables-are-available-windows 

There is a environment variable in MS Windows to get the user's home in an easy way.

**Comment by wenzeslaus** [comment]
@hellik What's wrong with [os.path.expanduser](https://docs.python.org/3/library/os.path.html#os.path.expanduser)?

> On Unix and Windows, return the argument with an initial component of ~ or ~user replaced by that user’s home directory.
>
> On Windows, USERPROFILE will be used if set, otherwise a combination of HOMEPATH and HOMEDRIVE will be used. An initial ~user is handled by stripping the last directory component from the created user path derived above.

File: gui/wxpython/startup/utils.py [hunk 2]
```diff
@@ -18,41 +18,48 @@
 
 import os
 import shutil
+import tempfile
+import sys
```
Comment (wenzeslaus) [inline gui/wxpython/startup/utils.py:22]: Seem to be unused.

File: gui/wxpython/startup/utils.py [hunk 3]
```diff
@@ -18,41 +18,48 @@
...
 
 
 def get_possible_database_path():
-    """Finds a path to what is possibly a GRASS Database. 
+    """Finds the directory to what is possibly a GRASS Database.
```
Comment (wenzeslaus) [inline gui/wxpython/startup/utils.py:26]: It is either "a path to what is" or "a directory which is".

File: gui/wxpython/startup/utils.py [hunk 4]
```diff
@@ -18,41 +18,48 @@
...
 
     Looks for directory named grassdata in the usual locations.
 
-    Returns the path as a string or None if nothing was found, so the
-    return value can be used to test if the directory was found.
+    Returns the path as a string or None if nothing was found, so the return 
+    value can be used to test if the directory was found.
```
Comment (wenzeslaus) [inline gui/wxpython/startup/utils.py:31]: Moving the word "return" is an unnecessary change and this adds white space at the end of line (which is not a good practice and now fortunately usually highlighted by Git).

File: gui/wxpython/startup/utils.py [hunk 5]
```diff
@@ -18,41 +18,48 @@
...
 
     Looks for directory named grassdata in the usual locations.
 
-    Returns the path as a string or None if nothing was found, so the
-    return value can be used to test if the directory was found.
+    Returns the path as a string or None if nothing was found, so the return 
+    value can be used to test if the directory was found.  
     """
-    home = os.path.expanduser('~')
+    home = os.path.expanduser('~')
```
Comment (wenzeslaus) [inline gui/wxpython/startup/utils.py:33]: Again, white space at the end of line. Flake8 would normally warn about that, but we have that warning still disabled in CI because of legacy code which importunately does have these problems (but let's not add more).

File: gui/wxpython/startup/utils.py [hunk 6]
```diff
@@ -18,41 +18,48 @@
...
-        os.path.join(home, "My Documents", "grassdata"),
+        os.path.join(home),
+        os.path.join(home, "Documents")
     ]
+    
     try:
         # here goes everything which has potential unicode issues
-        candidates.append(os.path.join(home, _("Documents"), "grassdata"))
-        candidates.append(os.path.join(home, _("My Documents"), "grassdata"))
+        candidates.append(os.path.join(home, _("Documents")))
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:50]: This adds Documents into the list of candidates twice in case _("Documents") == "Documents". So you could do:

```
tr_documents = os.path.join(home, _("Documents"))
if tr_documents not in candidates:
   candidates.append(tr_documents)
```

File: gui/wxpython/startup/utils.py [hunk 7]
```diff
@@ -18,41 +18,48 @@
...
+    # Find possible database path
+    for candidate in candidates:         
         if os.path.exists(candidate):
-            path = candidate
-            break  # get the first match
-    return path
+            for subdir in next(os.walk(candidate))[1]:
+                if 'grassdata' in subdir.lower():
+                    path = os.path.join(candidate, subdir)
+                    break
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:61]: break breaks only the inner loop, you need to return the path right away:
```
    for candidate in candidates:
        if os.path.exists(candidate):
            for subdir in next(os.walk(candidate))[1]:
                if 'grassdata' in subdir.lower():
                    return os.path.join(candidate, subdir)
    return None
```

then you don't need to define path variable at all.

File: gui/wxpython/startup/utils.py [hunk 8]
```diff
@@ -18,41 +18,48 @@
...
-    # ordered by preference and then likelihood
+    # ordered by platform preference      
+            
+    # Search independent "grassdata" directories      
     candidates = [
-        os.path.join(home, "grassdata"),
-        os.path.join(home, "Documents", "grassdata"),
-        os.path.join(home, "My Documents", "grassdata"),
+        os.path.join(home),
+        os.path.join(home, "Documents")
```
Comment (wenzeslaus) [inline gui/wxpython/startup/utils.py:45]: The point of trying "My Documents" is that we are not following the real API here of Windows or any other platform, so we are just adding this to the list of guesses. I put "My Documents" there because based on what I found at that time it seemed that it is a possible option (that there will be no "Documents" but only "My Documents"). This is does not seem to be the case for most recent Windows, but might be for some which are in use. I simply don't know, so I would prefer to keep "My Documents" there now, although I would like to see some analysis of this and a better solution in the (possibly distant) future.

**Comment by hellik** [comment]
> I put "My Documents" there because based on what I found at that time it seemed that it is a possible option (that there will be no "Documents" but only "My Documents"). This is does not seem to be the case for most recent Windows, but might be for some which are in use. I simply don't know, so I would prefer to keep "My Documents" there now, although I would like to see some analysis of this and a better solution in the (possibly distant) future.

AFAIK  my documents and all other variants, at least in more recent MS Windows versions, are just some kind of an alias for documents. Thus, you should be safe to use documents.

**Comment by ninsbl** [comment]
> AFAIK my documents and all other variants, at least in more recent MS Windows versions, are just some kind of an alias for documents. Thus, you should be safe to use documents.

Yes, that is the case here as well. What makes it far more worse however, is that - at least in my organisation - Windows home directories are linked to a Microsoft Sharepoint/OneDrive site (kind of google drive, dropbox, ...). There users will most likely experience more or less significant performance penalties because of syncing with online storage...

**Comment by mlennert** [comment]
Could you be a bit more explicit in your commit messages ? 'Improvements' and 'Improvements according to Vaclav' are just not very easy to interpret to get a quick info about the contents of the commit. ;-)

**Comment by lindakladivova** [comment]
> 
> 
> Could you be a bit more explicit in your commit messages ? 'Improvements' and 'Improvements according to Vaclav' are just not very easy to interpret to get a quick info about the contents of the commit. ;-)

Yes Moritz, I will be. Those commits are set up from several very small changes so I was not sure how to describe it. I will try to be more explicit next time.

**Comment by wenzeslaus** [comment]
> AFAIK my documents and all other variants, at least in more recent MS Windows versions, are just some kind of an alias for documents. Thus, you should be safe to use documents.

OK, let's than ignore any old or strange Windows and ignore also some Linux distros which had/has only the translated versions (based on my initial research). If we decide we need something like this in the future, we will probably need something more sophisticated anyway (there are Python packages and command line tools to deal with this properly).

@lindakladivova Please, remove all the `_("...")` candidates, this should simplify your code significantly.

File: gui/wxpython/startup/utils.py [hunk 9]
```diff
@@ -21,38 +21,47 @@
 
 
 def get_possible_database_path():
-    """Finds a path to what is possibly a GRASS Database. 
-
-    Looks for directory named grassdata in the usual locations.
-
-    Returns the path as a string or None if nothing was found, so the
-    return value can be used to test if the directory was found.
+    """Finds the directory for possible GRASS Database.
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:24]: The description is awkward. I suggest:

    """Looks for directory 'grassdata' (case-insensitive)
    in standard locations to detect existing GRASS Database.
    Returns the path as a string or None if nothing was found."""

File: gui/wxpython/startup/utils.py [hunk 10]
```diff
@@ -21,38 +21,47 @@
...
+    
+    Returns the path as a string or None. If nothing was found, 
+    the return value can be used to test if the directory was found.  
     """
-    home = os.path.expanduser('~')
+    home = os.path.expanduser('~')       
     # try some common directories for grassdata
-    # grassdata (lowercase) in home for Linux (first choice)
-    # Documents and My Documents for Windows
+    # case independent grassdata in home for Linux (first choice)
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:33]: most of the comment can be removed

**Comment by wenzeslaus** [comment]
> ... What makes it far more worse however, is that - at least in my organisation - Windows home directories are linked to a Microsoft Sharepoint/OneDrive site (kind of google drive, dropbox, ...). There users will most likely experience more or less significant performance penalties because of syncing with online storage...

@ninsbl Are you talking about using grassdata placed there or even just searching through that directory (one level only, two directories)?

I don't know what GRASS could do here. If you have a slow system or a bad setup, GRASS will be slow.

File: gui/wxpython/startup/utils.py [hunk 11]
```diff
@@ -21,38 +21,47 @@
...
-    return value can be used to test if the directory was found.
+    """Finds the directory for possible GRASS Database.
+    It automatically detects independent directory named grassdata 
+    in the usual locations. 
+    
+    Returns the path as a string or None. If nothing was found, 
+    the return value can be used to test if the directory was found.  
     """
-    home = os.path.expanduser('~')
+    home = os.path.expanduser('~')
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:31]: Again, you have additional whitespace at the end of line, please make sure to get rid of that before you commit. git diff will show you whitespace changes. Extra whitespace is bad practice and it clutters the PR changes.

Comment (neteler) [reply gui/wxpython/startup/utils.py]: @lindakladivova do you use a Python editor? If not, please consider to change - some of them are able to remove trailing white space automagically.

File: gui/wxpython/startup/utils.py [hunk 12]
```diff
@@ -21,38 +21,30 @@
...
-
-    Returns the path as a string or None if nothing was found, so the
-    return value can be used to test if the directory was found.
+    """Looks for directory 'grassdata' (case-insensitive) in standard 
+    locations to detect existing GRASS Database.
+    
+    Returns the path as a string or None if nothing was found.
     """
-    home = os.path.expanduser('~')
+    home = os.path.expanduser('~')
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:29]: whitespace...

File: gui/wxpython/startup/utils.py [hunk 13]
```diff
@@ -21,38 +21,30 @@
...
     # try some common directories for grassdata
-    # grassdata (lowercase) in home for Linux (first choice)
-    # Documents and My Documents for Windows
-    # potential translations (old Windows and some Linux)
-    # but ~ and ~/Documents should cover most of the cases
-    # ordered by preference and then likelihood
+    # case independent grassdata in home for Linux (first choice)
+    # Documents for Windows
+            
+    # Independent "grassdata" directories
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:34]: you mean "case independent grassdata directories"?

File: gui/wxpython/startup/utils.py [hunk 14]
```diff
@@ -21,38 +21,30 @@
...
-        candidates.append(os.path.join(home, _("Documents"), "grassdata"))
-        candidates.append(os.path.join(home, _("My Documents"), "grassdata"))
-    except UnicodeDecodeError:
-        # just ignore the errors if it doesn't work
-        pass
-    path = None
-    for candidate in candidates:
+        home,
+        os.path.join(home, "Documents"),
+        os.path.join(home, "My Documents")
```
Comment (petrasovaa) [inline gui/wxpython/startup/utils.py:38]: Based on @hellik's comment, we don't need My Documents anymore.

File: gui/wxpython/startup/utils.py (hunk 11 above)
Comment (lindakladivova) [reply gui/wxpython/startup/utils.py]: I have switched on blank spaces in Spyder, so now hopefully, I will be able to see trailing white spaces.
